docs/Garbage Collection.md

```diff
@@ -35,3 +41,282 @@ by calling `JS_GC` or `JS_MaybeGC`.
+No such snapshot is physically taken — doing so would require a full nonincremental mark!
+
+The implementation of the snapshot-at-the-beginning barrier is as
+follows.
+The barrier fires whenever the mutator is about to overwrite a location
+that holds a GC pointer.
+The barrier marks the pointed-to object black, and pushes any outgoing
+edges onto the mark stack.
+The key observation is that the only way an object can get 'lost' and
+not marked is if all pointers to the object are overwritten.
+Thus, if whenever we're about to overwrite a pointer to an object we
+mark the object black first, then no objects can get 'lost'.
+
+Now consider allocations.
+A newly allocated object wasn't present at the beginning of the GC, so
+the snapshot-at-the-beginning barrier doesn't do it any good.
+But we do need to make sure the object doesn't get collected if it's
+live.
+This is easily done by marking new objects immediately upon allocation
+during an incremental GC, thus the name **allocate-black**.
+
+### The SpiderMonkey incremental read barrier ###
+
+The textbook version of incremental GC has only a write barrier.
+SpiderMonkey also needs a read barrier for weak references.
+This is because weak references have a similar problem as is solved by
+the incremental write barrier above, but cannot use the same solution.
+
+By definition, a weak reference should not hold a target live by itself.
+That means that we can't trace weak references when we encounter them
+during GC graph traversal.
+Instead, we append them to a list.
+After marking everything live we scan through the list to null out
+references to garbage.
+
+But this also means that pre-barriers are inadequate to prevent
+incremental GC from missing edges due to graph mutation.
+Live cells will be marked either because they're reachable or because
+they were marked by a pre-barrier before being removed from the graph.
+But reachable weak references aren't traced!
+So you could read one out and write it behind the frontier and no pre-barriered removal will mark it (it's not being removed, just read).
+So we barrier the read instead; if you ever read a weak reference in the
+middle of an ongoing incremental GC, then the target of the weak ref
+will be marked.
+
+### Implementation details ###
+
+Write barriers have a runtime cost, so SpiderMonkey tries to skip them
+when an incremental GC cycle is not active.
+Each Zone has a flag `needsIncrementalBarrier()` that indicates whether
+barriers are required.
+
+For each type `T` such that fields of type `T*` need a write barrier,
+there is a function `T::writeBarrierPre(old)`.
+For example, fields of type `JSObject*` need a write barrier, so there
+is a function `JSObject::writeBarrierPre(JSObject* old)`.
+If `zone->needsIncrementalBarrier()`, then `writeBarrierPre()` marks
+`old`.
+That's it.
+
+The class `Heap<T>` is provided to make it easy to invoke write
+barriers.
+A `Heap<T>` encapsulates a `T*` and invokes the write barrier whenever
```

Comment:
Heap<T> encapsulates a `T`. Generally things are Heap<T*> for engine types like JSObject.